mots: create initial prototype (bug 1733956)

zzzeid opened the pull request:
A helper Python library to facilitate in-tree module ownership system integration.

zzzeid commented on src/mots/bmo.py:
```diff
@@ -0,0 +1,108 @@
+        else:
+            return None
+        return result["users"][0]
+
+    # NOTES:
+    # How to store people?
+    # 64523: glob  -> if nick is available
+    # 65182: Jane Smith  -> if nick is not available, use provided name.
+    # OR
+    # - Masayuki Nakano bmo:19563
```
```suggestion
```

zzzeid commented on src/mots/bmo.py:
```diff
@@ -0,0 +1,108 @@
+"""Module that provides helpers to interact with the Bugzilla API."""
+
+# TODO: this should be abstracted so that it can be hot-swappable with any API that can
+# be searched.
+
```
```suggestion

```

cgsheeh commented on src/mots/bmo.py:
```diff
@@ -0,0 +1,108 @@
+    def __init__(self, token: str = None, base_url: str = DEFAULT_BASE_URL):
+        if not token:
+            token = os.getenv("BUGZILLA_API_KEY", "")
+            if not token:
+                token = input("Enter BMO Token: ")
+            if not token:
+                raise Exception()
```
I'm -1 on using `input` here. If we really need to use it, it should probably be a level above in the CLI.

We should add some more useful information to that exception, and maybe make it a `ValueError`.

cgsheeh commented on src/mots/bmo.py:
```diff
@@ -0,0 +1,108 @@
+
+    def get_user(self, email: str, rate_limit_delay: float = 0.3):
+        """Get user data based on provided email address."""
+        time.sleep(rate_limit_delay)
```
I don't think sleeping before making any requests to BMO is the right approach. We could record the timestamp of the request as `self.last_request` and wait if two requests are within the rate limit delay instead. What do you think?

zzzeid replied:
This is actually obsolete now, since everything is done in a single API request, I will update this code and remove the rate limits.

cgsheeh commented on src/mots/bmo.py:
```diff
@@ -0,0 +1,108 @@
+        )
+        if response.status_code != 200:
+            logger.error(f"Error searching {email}: {response.status_code}")
+            return
```
Should this raise?

cgsheeh commented on src/mots/cli.py:
```diff
@@ -0,0 +1,240 @@
+            modules = result[path]
+            module_names = ",".join([m.machine_name for m in modules])
+            owners = ",".join([",".join(m.owners) for m in modules])
+            sys.stdout.write(f"{path}:{module_names}:{owners}\n")
```
Any reason for not using `print` here as we do elsewhere?

cgsheeh commented on src/mots/directory.py:
```diff
@@ -0,0 +1,254 @@
+        people = list(people)
+        self.people = []
+        self.by_bmo_id = {}
+        for i in range(len(people)):
```
`for i, person in enumerate(people):`

cgsheeh reviewed (changes requested):
This was my first pass. I'll jump in a bit deeper tomorrow.

zzzeid commented:
> This was my first pass. I'll jump in a bit deeper tomorrow.

Thanks for the feedback @cgsheeh -- definitely a few remnants from the early approaches left in here, which you've commented on (e.g. the use of `input` and some of the issues in `cli` as well as generic exceptions.) I'll address the feedback tomorrow but do expect a few more rough edges.

mitchhentges commented on src/mots/bmo.py:
```diff
@@ -0,0 +1,108 @@
+    def get_user_by_id(self, id_: str = "", rate_limit_delay: float = 0.3):
+        """Get user data by BMO ID."""
+        time.sleep(rate_limit_delay)
+        fields = ["real_name", "nick", "name", "id"]
```
We probably want to return the same `user` data structure, regardless of which `get_user_by_*` is being called.
Perhaps we should reuse `fields`?

zzzeid replied:
This is also not being used any more. Will delete it.

mitchhentges commented on src/mots/cli.py:
```diff
@@ -0,0 +1,240 @@
+            sys.stdout.write(f"{path}:{module_names}:{owners}\n")
+
+    @staticmethod
+    def add(args):
+        """Prompt user to add a new module."""
+        params = {
+            "machine_name": input("Enter machine name of new module: "),
+            "name": input("Enter a human readable name: "),
+            "owners": get_list_input("Enter a comma separated list of owners"),
+            "peers": get_list_input("Enter a comma separated list of peers"),
+            "includes": get_list_input(
+                "Enter a comma separated list of paths to include"
+            ),
+            "excludes": get_list_input(
+                "Enter a comma separated list of paths to exclude"
+            ),
+        }
+        parent = input("Enter a machine name of the parent module (optional): ") or None
+        file_config = FileConfig(Path(args.path))
+        module.add(params, file_config, parent=parent, write=True)
```
In regular use, are you preferring that users make changes to the yaml directly, or via the CLI?

zzzeid replied:
It's really up to whoever is changing the file to make their own decisions. Either edit it manually, push, and let Lando/Phabricator do their thing and show the automatically parsed changes, or use `mots` CLI. There is no preference on my end, since `mots clean` will run regardless before things land.

mitchhentges commented on src/mots/module.py:
```diff
@@ -0,0 +1,356 @@
+                module["submodules"].append(serialized)
+                break
+    else:
+        modules.append(serialized)
```
Is there any validation we can/should do to provided data?

zzzeid replied:
Yes, we could trigger `clean` or `validate` after this. I think I will rewrite this as part of improving overall general and schema validation

zzzeid replied:
Filed [a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1753716) for this.

mitchhentges commented on src/mots/config.py:
```diff
@@ -0,0 +1,56 @@
+        """Initialize the configuration with provided config file path."""
+        if not path.exists() and path.is_file():
+            raise ValueError(f"{path} does not exist or is not a file.")
+        self.path = path
+        self.repo_path = path.parent
+        self.config = None
```
It might be worth following the "internal members behind `_` prefixes" convention, to tighten API surfaces in advance.

mitchhentges commented on src/mots/directory.py:
```diff
@@ -0,0 +1,254 @@
+            for path in self.repo_path.glob("**/*"):
+                if path.relative_to(self.repo_path).parts[0] in (".hg", ".git"):
+                    continue
+                self.index[path] = list()
```
Is it OK that this will include potentially-large ignored directories, like `obj-*`?

zzzeid replied:
Filed [a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1753713) for this.

mitchhentges commented on src/mots/directory.py:
```diff
@@ -0,0 +1,254 @@
+class Person:
+    """A class representing a person."""
+
+    bmo_id: int = None
```
If you'd prefer to keep them non-optional (which is good IMO), I'd recommend doing `assert bmo_data` instead of `if bmo_data` in `__post_init__()`, so that if `bmo_data` is `None` then an error will be raised, rather than having an empty `Person` object.

mitchhentges commented on src/mots/directory.py:
```diff
@@ -0,0 +1,254 @@
+        logger.debug(f"Initializing people directory with {len(people)} people...")
+        if query_bmo:
+            bmo_client = BMOClient()
+            bmo_data = bmo_client.get_users_by_ids([p["bmo_id"] for p in people])
```
For better test ergonomics, I'd recommend moving "work that touches the outside" (such as HTTP requests) to a separate function.
For example, turn `People` into a `@dataclass`, but give it a `def from_bmo()` function that production code can use.

mitchhentges commented on src/mots/cli.py:
```diff
@@ -0,0 +1,240 @@
+    def add(args):
+        """Prompt user to add a new module."""
+        params = {
+            "machine_name": input("Enter machine name of new module: "),
```
This prompt may be hard to understand for people not familiar with the nomenclature: "what's the machine name? _My_ machine's name?" 😉

mitchhentges commented:
Pretty exciting so far, keep it up :)

cgsheeh reviewed (commented):
I saw a few more small nits/things that stood out to me. I'm going to hold off on a more thorough review until some of the deprecated code is removed and we have our meeting next week. :)

zzzeid commented on Makefile:
```diff
@@ -0,0 +1,101 @@
+.PHONY: format
+format:
+	$(PYTHON) -m black src/mots
+	$(PYTHON) -m black tests
```
```suggestion
	$(PYTHON) -m black tests
	$(PYTHON) -m black documentation
```

zzzeid commented on documentation/index.rst:
```diff
@@ -0,0 +1,150 @@
+
+commands:
+  {init,module,validate,clean,query}
+    init                initialize mots configuration in repo
```
In this section, this is the output of `--help`, it's not generated. But I think it would be good to include this documentation somewhere in some format -- this is not the ideal format or way to include it though. I could reformat it and factor out some of the repetitive output to make it easier on the eyes.

zzzeid commented on src/mots/bmo.py:
```diff
@@ -0,0 +1,108 @@
+        )
+        return response
+
+    def get_user(self, email: str, rate_limit_delay: float = 0.3):
+        """Get user data based on provided email address."""
+        time.sleep(rate_limit_delay)
+        fields = ["real_name", "nick", "email", "name", "id"]
+        response = self._get(
+            "user", {"names": email, "include_fields": ",".join(fields)}
+        )
+        if response.status_code != 200:
+            logger.error(f"Error searching {email}: {response.status_code}")
+            return
+
+        # TODO: handle case when there is more than one match.
+        # Since possibly users can sneak in their email address in another field?
+        # Although we should validate the "email" field in BMO response + check for
+        # verification.
+
+        result = response.json()
+        return result["users"][0]
```
```suggestion
```

zzzeid commented on src/mots/bmo.py:
```diff
@@ -0,0 +1,108 @@
+
+        result = response.json()
+        return result["users"][0]
+
+    def get_user_by_id(self, id_: str = "", rate_limit_delay: float = 0.3):
+        """Get user data by BMO ID."""
+        time.sleep(rate_limit_delay)
+        fields = ["real_name", "nick", "name", "id"]
+        response = self._get("user", {"ids": id_, "include_fields": ",".join(fields)})
+        if response.status_code != 200:
+            logger.error(f"Error searching {id_}: {response.status_code}")
+            return
+
+        result = response.json()
+        return result["users"][0]
```
```suggestion
```

zzzeid commented on src/mots/module.py:
```diff
@@ -0,0 +1,356 @@
+        print(modules_by_name[module])
+
+
+def extract_people(module):
```
Yes, this was one of the scaffolding helper methods I used when first generating the yaml file, will clean up.

zzzeid commented on src/mots/module.py:
```diff
@@ -0,0 +1,356 @@
+
+        if self.submodules:
+            for submodule in self.submodules:
+                return submodule.validate(errors=errors)
```
Nice catch. This is meant to continue iterating on all submodules, but the test was only testing with one submodule. Fixing the test + method.

zzzeid commented on src/mots/module.py:
```diff
@@ -0,0 +1,356 @@
+
+    :raises ValueError: when no module matches the given machine name
+    """
+    modules_by_name = {}
+    for _module in modules:
+        modules_by_name[_module["machine_name"]] = _module
+    if module not in modules_by_name:
+        raise ValueError(f"{module} not found.")
+    else:
+        print(modules_by_name[module])
```
Makes sense. I think the current implementation is very rudimentary anyway, and will likely morph into something different in the future. Will fix tho.

zzzeid commented on src/mots/module.py:
```diff
@@ -0,0 +1,356 @@
+    :param file_config: an instance of :class:`FileConfig`
+    :param write: if set to `True`, writes changes to disk.
+    """
+    from mots.directory import Directory
```
This is done here to prevent circular import. There might be an alternate way to structure the imports but as these two modules are currently laid out, this import needs to be here.

cgsheeh commented on src/mots/config.py:
```diff
@@ -0,0 +1,56 @@
+            # File does not exist, create it.
+            now = datetime.now().isoformat()
+            self.config = {
+                "repo": str(Path(self.path).resolve().parts[-2]),
```
Is `-2` specific to `./mots.yaml` here? Could a different path be passed that breaks this, ie via the `--path` CLI argument? Maybe it's fine since the `path` arg to `init` is for the repo itself, but just want to put it on your radar.

zzzeid replied:
So the answer is yes, it is specific to this path. Since the path to the config file is assumed to be in the top level directory of the repo, this should never cause any problems. It might be good to add some more explicit checks to make sure that `mots.yaml` is always at the top level, but basically if this is not the case, then other things will break. Agreed that this needs to evolve a little more though -- maybe once I add more support for other source control awareness (e.g. `hg files`).

cgsheeh replied:
Makes sense. I won't block the review on this, but adding a TODO comment here would be nice.

zzzeid replied:
Filed [a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1753714) for this.

cgsheeh reviewed (changes requested):
This was a deeper dive. Most of the core logic seems sound to me! :+1: 

Meta note: please use proper variable names in loops, ie no `m`, `p`, `sm` etc. There were many instances so I stopped commenting on them all. Same with `range(len(obj))` vs `enumerate`.

mitchhentges commented on src/mots/bmo.py:
```diff
@@ -0,0 +1,70 @@
+        )
+        return response
+
+    def get_users_by_ids(self, ids: list[str]):
```
[Looks like Python 3.9](https://docs.python.org/3/whatsnew/3.9.html#type-hinting-generics-in-standard-collections).

zzzeid replied:
mots requires Python 3.7, and the [future import above](https://github.com/mozilla-conduit/mots/pull/1/files#diff-a0f469c2494151363815be0c78dc76e67e8873dc8488e4004396f1bc13ab73aaR7) will take care of this specific issue.

mitchhentges replied:
[Ah, you're right](https://www.python.org/dev/peps/pep-0585/#backwards-compatibility).

cgsheeh commented on src/mots/cli.py:
```diff
@@ -0,0 +1,238 @@
+    directory = Directory(file_config)
+    directory.load()
+    result = directory.query(*args.paths)
+    for path in result.path_map:
+        modules = result.path_map[path]
```
`for path, modules in result.path_map.items():`

cgsheeh commented on src/mots/config.py:
```diff
@@ -0,0 +1,206 @@
+
+    # Count machine name repetitions in a defaultdict.
+    machine_names = defaultdict(int)
+    for m in modules:
+        machine_names[m["machine_name"]] += 1
+        if "submodules" in m and m["submodules"]:
+            for sm in m["submodules"]:
+                machine_names[sm["machine_name"]] += 1
```
Variable names: `m` -> `module`, `sm` -> `submodule`

cgsheeh reviewed (changes requested):
Looks good, the main thing remaining here is cleanup of the `try/except KeyError` block and followup on the `-2` thing.

I pulled this PR down and ran the CLI a few times. I noticed a few issues: first the representations printed to the screen are Python text representations (like `[defaultdict()...]` etc). If we don't want to block this on landing, perhaps we could `pprint.pprint` instead of `print` as a quick solution.

Next, when I run `mots query` I see a Bugzilla error. Is this expected/can we make this better?

```shell
❯ BUGZILLA_API_KEY=<redacted> mots query src/
2022-02-03 11:46:17,698 directory  WARNING  No bugzilla data was provided for user 633708
src/:main
```  

There is also a lot of unrelated spew when running `mots` without specifying the proper subcommand/arguments. See below, where `mots --help` prints what I would expect for `mots`, and `mots module` without a subcommand also prints the help for every command:

```shell
❯ mots --help
usage: mots [-h] [--debug] {init,module,validate,clean,query,export} ...

main command line interface for mots

optional arguments:
  -h, --help            show this help message and exit
  --debug               enable debug output

commands:
  {init,module,validate,clean,query,export}
    init                initialize mots configuration in repo
    module              module_operations
    validate            validate mots config for current repo
    clean               clean mots config for current repo
    query               query the module directory
    export              export the module directory
❯ mots
usage: mots [-h] [--debug] {init,module,validate,clean,query,export} ...

main command line interface for mots

optional arguments:
  -h, --help            show this help message and exit
  --debug               enable debug output

commands:
  {init,module,validate,clean,query,export}
    init                initialize mots configuration in repo
    module              module_operations
    validate            validate mots config for current repo
    clean               clean mots config for current repo
    query               query the module directory
    export              export the module directory

usage: mots init [-h] [--path PATH]

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  the path of the repo to initialize

usage: mots module [-h] [--path PATH] {list,add,show} ...

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  the path of the repo config file

module:
  {list,add,show}
    list                list all modules
    add                 add a new module
    show                show a module

usage: mots validate [-h] [--path PATH] [--repo-path REPO_PATH]

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  the path of the repo config file
  --repo-path REPO_PATH, -r REPO_PATH
                        the path of the repo

usage: mots clean [-h] [--path PATH] [--repo-path REPO_PATH]

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  the path of the repo config file
  --repo-path REPO_PATH, -r REPO_PATH
                        the path of the repo

usage: mots query [-h] [--path PATH] paths [paths ...]

positional arguments:
  paths                 a list of paths to query

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  the path of the repo config file

usage: mots export [-h] [--path PATH] [--format FORMAT]

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  the path of the repo config file
  --format FORMAT, -f FORMAT
                        the format of exported data
❯ mots module
usage: mots [-h] [--debug] {init,module,validate,clean,query,export} ...

main command line interface for mots

optional arguments:
  -h, --help            show this help message and exit
  --debug               enable debug output

commands:
  {init,module,validate,clean,query,export}
    init                initialize mots configuration in repo
    module              module_operations
    validate            validate mots config for current repo
    clean               clean mots config for current repo
    query               query the module directory
    export              export the module directory

usage: mots init [-h] [--path PATH]

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  the path of the repo to initialize

usage: mots module [-h] [--path PATH] {list,add,show} ...

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  the path of the repo config file

module:
  {list,add,show}
    list                list all modules
    add                 add a new module
    show                show a module

usage: mots validate [-h] [--path PATH] [--repo-path REPO_PATH]

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  the path of the repo config file
  --repo-path REPO_PATH, -r REPO_PATH
                        the path of the repo

usage: mots clean [-h] [--path PATH] [--repo-path REPO_PATH]

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  the path of the repo config file
  --repo-path REPO_PATH, -r REPO_PATH
                        the path of the repo

usage: mots query [-h] [--path PATH] paths [paths ...]

positional arguments:
  paths                 a list of paths to query

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  the path of the repo config file

usage: mots export [-h] [--path PATH] [--format FORMAT]

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  the path of the repo config file
  --format FORMAT, -f FORMAT
                        the format of exported data

```

zzzeid commented:
> Looks good, the main thing remaining here is cleanup of the `try/except KeyError` block and followup on the `-2` thing.

The try/except thing is fixed now.

> I pulled this PR down and ran the CLI a few times. I noticed a few issues: first the representations printed to the screen are Python text representations (like `[defaultdict()...]` etc). If we don't want to block this on landing, perhaps we could `pprint.pprint` instead of `print` as a quick solution.

Fixed.


> Next, when I run `mots query` I see a Bugzilla error. Is this expected/can we make this better?

Fixed.

> There is also a lot of unrelated spew when running `mots` without specifying the proper subcommand/arguments. See below, where `mots --help` prints what I would expect for `mots`, and `mots module` without a subcommand also prints the help for every command:

Fixed. There was originally a for loop that iterated over all subparsers and printed the help. This (may have) made sense when there were only 2 commands but now it does not. I think I originally added for debugging, so nice catch!

zzzeid commented on requirements.in:
```diff
@@ -0,0 +1,11 @@
+black==21.7b0
```
Going to hold off on that for now, but will explore this idea for the future.

zzzeid replied:
> TL;DR: I'm ok with leaving this as-is, but I want to make sure that I'm properly communicating why pip-compile isn't sufficient for the long-term.

Noted, and I am aware of the limitations of pip-compile that you mentioned (MozPhab has a similar but slightly more complex situation, and there I've solved that with simple docker containers to generate those files for each platform). I don't necessarily see this issue as needing another library and additional config files for a long term solution, but it's certainly something to consider.

Keep in mind that the requirements.txt file is mostly needed for CircleCI purposes, and also for local development. `mots` itself only has two requirements which are included in `setup.cfg` and does not require any lock files or anything like that. For CircleCI we always know ahead of time which platform we'll be running tests on.

zzzeid replied:
Another minor comment here is that the only requirement that is python-version-specific for packaging `mots`, is `importlib-resources`. It's definitely not an unimportant part of the code (since it will be used when exporting to rst) but I'd like to find a cleaner solution to fetch templates in the future, so this conditional dependency may be removed in the future.

zzzeid replied:
Filed a [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1753756) for this. The requirements files will need to be fixed before merging/deploying this since tests will actually fail on Python 3.7 as things stand now.

mitchhentges commented on src/mots/config.py:
```diff
@@ -0,0 +1,214 @@
+                    ]
+
+        # Do the same for submodules.
+        # TODO: this should be refactored.
```
Let's move this "`TODO`" to a bug.
If you wouldn't mind, can you pop a "find all" for "`TODO`" and move 'em all?

zzzeid replied:
[bug filed](https://bugzilla.mozilla.org/show_bug.cgi?id=1753753). Will replace the rest of TODOs with bugs, where needed.

mitchhentges replied:
Note that there's still a couple `TODO`s in the tests, if you want to snatch those up too.

mitchhentges commented on src/mots/directory.py:
```diff
@@ -0,0 +1,216 @@
+class QueryResult:
+    """Helper class to simplify query result interpretation."""
+
+    paths: list = None
+    path_map: dict = None
+    rejected_paths: list = None
+    modules: list = None
+    owners: list = None
+    peers: list = None
```
I don't think that these variables are used.
I'm guessing that they're here to imply typings for `data_keys`, but they might be better represented as docs - currently they're misleading due to not being written/accessed.

Also, are the `None` values necessary?

mitchhentges commented on src/mots/directory.py:
```diff
@@ -0,0 +1,216 @@
+    """A class representing a person."""
+
+    bmo_id: int
+    name: str = ""
+    info: str = ""
+    nick: str = ""
```
Are `name,` `info` and `nick` optional? If not, the default value here may be redundant - especially since the default value of `""` is provided down in the `bmo_data.get("<val>", "")` calls.

mitchhentges commented on src/mots/directory.py:
```diff
@@ -0,0 +1,216 @@
+
+    def __post_init__(self, bmo_data):
+        """Refresh BMO data from BMO API."""
+        if bmo_data:
```
This `if` statement is a smell that we have two different use paths here:
1. Where we just init the dataclass the "normal" way: `Person(0, "name", "info", "nick")`
2. Where we're parsing raw-ish output Bugzilla: `Person(0, bmo_data)`.

I'd recommend removing `__post_init__()` and instead adding a static `parse()` method, [kinda like phabricator emails](https://github.com/mozilla-conduit/phabricator-emails/blob/master/phabricatoremails/render/events/phabricator.py#L39-L51).

This gives you two benefits:
1. You're no longer setting `bmo_id` twice
2. The API of this class is easier to use, because you use either the constructor _or_ the dedicated `Person.parse()` static function depending on the use case you're using.

mitchhentges commented on src/mots/directory.py:
```diff
@@ -0,0 +1,216 @@
+            self.name = parsed_real_name["name"]
+            self.info = parsed_real_name["info"]
+
+    def __hash__(self):
+        """Return a unique identifier for this person."""
+        return int(self.bmo_id)
```
Smart :)
FWIW, `self.bmo_id` should already be an int, right? If `bmo_data["id"]` is a string, we should parse it accordingly so that the data is clean by the time it's represented by this data object.

mitchhentges commented on src/mots/module.py:
```diff
@@ -0,0 +1,197 @@
+
+logger = logging.getLogger(__name__)
+
+print = pprint.PrettyPrinter().pprint
```
Nit:
```
from pprint import pprint
...
print = pprint
```
That should be functionally equivalent to `pprint.PrettyPrinter().pprint`, right?

zzzeid replied:
I suppose we could even do

```python
from pprint import pprint as print
```

mitchhentges commented on src/mots/module.py:
```diff
@@ -0,0 +1,197 @@
+        from the directory index
+    """
+
+    def __repr__(self):
+        """Return a human readable name for the instance."""
+        return f"<Module: {self.machine_name}>"
```
I like how usable you made your interactive Python API with nice helpers like this, nice work :+1:

mitchhentges commented on src/mots/module.py:
```diff
@@ -0,0 +1,197 @@
+        self,
+        machine_name: str,
+        repo_path: str,
+        name: str = None,
+        description: str = None,
```
Small nit:
Having an `Optional[str]` has essentially three states:
* `None`
* `""`
* `"<contents>"`

For most intents and purposes, `None` == `""`, and having to track _both_ cases for validation/printing/etc can be a pain.
It might be worth considering leaning towards using `""` to represent "unset string properties" instead of `None` to simplify related logic.

zzzeid replied:
Fixed those two, and also the type hints were incorrect for `includes` and `excludes` below, which are also fixed now.

mitchhentges commented on src/mots/utils.py:
```diff
@@ -0,0 +1,59 @@
+    return [e.strip() for e in user_input if e]
+
+
+def parse_user_string(string):
```
I don't think this is used

zzzeid replied:
Nice catch. This lingered on from the initial migration of the wiki page; will remove.

mitchhentges commented on src/mots/utils.py:
```diff
@@ -0,0 +1,59 @@
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+"""Utility helper functions."""
```
This utilities can probably all be moved closer to their associated domains for better discoverability:

* `generate_machine_readable_name`: This might be better being beside the `class Module` definition.
* `get_list_input`: Chuck this in `cli.py` IMHO
* `parse_real_name`: This can be beside `class Person`

zzzeid replied:
I do like being able to separate and test these small methods that do not rely on any external classes in their own module (i.e. `utils`). I can see why they could be moved to a different module if they had any dependencies, but they're basically text parsers that may be used by any other module -  thus I think they're fine to stay here.

mitchhentges reviewed (changes requested):
Looks pretty good, got a couple nits and a couple needed cleanups (specifically the `TODO`s).

To make sure this is tracked with the PR: as mentioned on Slack, it _may_ be worth using a single BMO identifier in the `mots.yaml`, then populating the rest from BMO at output-generation-time. This would allow reducing `mots.yaml` churn and removing the need for a CLI used by end developers.
However, due to the significant changes required by the endeavour and the stage of this project, it's not a blocking issue - it may be worth reconsidering in the future though.

-----

TL;DR: Nice work, looking forward to deploying this "for real" in-tree :)

zzzeid commented on .circleci/config.yml:
```diff
@@ -0,0 +1,36 @@
+  version: 2
+  test:
+    jobs:
+      - test_3_9
```
Jobs added and passing for Pythons 3.7 to 3.10.

zzzeid commented on src/mots/bmo.py:
```diff
@@ -0,0 +1,65 @@
+logger = logging.getLogger(__name__)
+
+DEFAULT_BASE_URL = "https://bugzilla.mozilla.org/rest"
+USER_AGENT = "mots"  # TODO: improve this and include version.
```
[bug filed](https://bugzilla.mozilla.org/show_bug.cgi?id=1753752)

zzzeid commented on .circleci/config.yml:
```diff
@@ -0,0 +1,51 @@
+            set -e
+            sudo apt-get install python3-venv
+            mkdir -p /tmp/test-reports
+            make dev-env PY=python3.9
```
```suggestion
            make dev-env
```

cgsheeh commented on requirements/base.in:
```diff
@@ -0,0 +1,10 @@
+black==22.1.0
+build==0.7.0
```
Missing `docker-compose`?

zzzeid replied:
`docker-compose` depends on `setuptools` -- which will generate some warnings in the requirements files. I personally use a system-wide docker-compose. Can we put a pin in this? (pun NOT intended)

mitchhentges replied:
Note that if the warning is the same warning I'm thinking of around `--allow-unsafe`, the [`pip-compile` folks are considering deprecating it](https://github.com/jazzband/pip-tools/issues/989).

cgsheeh reviewed (approved):
Well done!

mitchhentges commented on src/mots/cli.py:
```diff
@@ -0,0 +1,234 @@
+
+def init(args: argparse.Namespace) -> None:
+    """Call `file_config.init` with correct arguments."""
+    file_config = config.FileConfig(Path(args.path))
```
Sure, if you'd prefer to keep just `args` that's ok with me.

> of which at least one will be somewhat redundant (e.g. the example you provided applied to init)

FWIW, I couldn't find any CLI functions that would have `file_config` be rudundant.

zzzeid replied:
> FWIW, I couldn't find any CLI functions that would have file_config be rudundant.

I should have specified which argument I was referring to. If we take this approach, then in some cases, the `args` argument will become redundant or unused -- e.g. in `init`, `ls`, `clean`. As mentioned, this is in the context of keeping all the method signatures the same.

zzzeid commented on Makefile:
```diff
@@ -0,0 +1,110 @@
+	source ./.mots-env/bin/activate
+	python -m pip install --upgrade pip
+	python -m pip install pip-tools
+	ls -lla requirements
```
```suggestion
```

mitchhentges reviewed (approved):
Nice work! :clap:
